### PR TITLE
NEW: Feature Extraction for Deep Feature Based Registration

### PR DESCRIPTION
**Pull request by ruqayya**
It contains a function for computing CNN-based multiscale features for an image pair.

### CODE REVIEW

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import cv2
+import numpy as np
+import torch
+import torchvision
+from torchvision.models._utils import IntermediateLayerGetter
+
+
+class DFBR:
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:8]: How about

```suggestion
class DFBRegistration:
```
?

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    r"""Deep Feature based Registration
+
+    This class implements a CNN feature based registration,
+    as proposed in a paper titled `Deep Feature based Cross-slide Registration
+    <https://arxiv.org/pdf/2202.09971.pdf>`_.
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:13]: Are you able to add any metrics?

File: tests/test_wsi_registration.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    fixed_img = np.repeat(np.expand_dims(fixed_img, axis=2), 1, axis=2)
+    moving_img = np.repeat(np.expand_dims(moving_img, axis=2), 1, axis=2)
+    with pytest.raises(
+        ValueError, match=r".*The input images are expected to have 3 channels.*"
+    ):
+        _ = df.extract_features(fixed_img, moving_img)
+
+    fixed_img = np.repeat(fixed_img, 3, axis=2)
+    moving_img = np.repeat(moving_img, 3, axis=2)
+    _ = df.extract_features(fixed_img, moving_img)
```
Comment (shaneahmed) [inline tests/test_wsi_registration.py:34]: you should use assert for any checks. e.g., you can use some metrics.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+        if fixed_img.shape[2] != 3 or moving_img.shape[2] != 3:
+            raise ValueError("The input images are expected to have 3 channels.")
+
+        self.Xscale = 1.0 * np.array(fixed_img.shape[:2]) / self.patch_size
+        self.Yscale = 1.0 * np.array(moving_img.shape[:2]) / self.patch_size
+        fixed_cnn = cv2.resize(fixed_img, self.patch_size)
+        moving_cnn = cv2.resize(moving_img, self.patch_size)
+
+        fixed_cnn = fixed_cnn / 255
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:57]:
```suggestion
        fixed_cnn = fixed_cnn / 255.
```

Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:57]: If it is float then use 255.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        moving_cnn = moving_cnn / 255
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:58]:
```suggestion
        moving_cnn = moving_cnn / 255.
```
same as above.

**Review by shaneahmed — CHANGES_REQUESTED** [review]
Please can you make the requested changes?

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 2 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py:13]: What do you mean by metrics here? Can you please clarify?

Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:13]: Accuracy, RMSE, TRE, ...

Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py:13]: These metrics are not needed at this point for this PR. However, they can be computed against the manual landmark points once we have computed the transformation parameters. I will add a separate function for metrics later.

File: tests/test_wsi_registration.py (hunk 3 above)
Comment (ruqayya) [reply tests/test_wsi_registration.py:34]: This feature extraction step is followed by feature mapping and matching steps. Evaluation metrics will be used at later stages.

Comment (ruqayya) [reply tests/test_wsi_registration.py:34]: Please check: https://github.com/TissueImageAnalytics/tiatoolbox/blob/edc7e7f770dc66d154a5ddb26d87b4fed7789d84/tests/test_wsi_registration.py#L46-L55

File: tests/test_wsi_registration.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
```
Comment (shaneahmed) [inline tests/test_wsi_registration.py:34]:
```suggestion
```

File: tests/test_wsi_registration.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import PIL.Image as Image
+import pytest
+
+from tiatoolbox.tools.registration.wsi_registration import DFBRegistrtation
+
+
+def test_extract_features():
+    """Test for CNN based feature extraction function."""
+    main_url = "https://tiatoolbox.dcs.warwick.ac.uk/testdata/registration/"
+    urllib.request.urlretrieve(main_url + "HE_1_level8_gray.png", "sample.png")
```
Comment (shaneahmed) [inline tests/test_wsi_registration.py:13]: Please use `tmp_path` and `conftest.py`. Please see links below for examples.

https://github.com/TissueImageAnalytics/tiatoolbox/blob/af7319ef633e2f8b315b0949ed343f67d38bab0a/tests/conftest.py#L52

https://github.com/TissueImageAnalytics/tiatoolbox/blob/af7319ef633e2f8b315b0949ed343f67d38bab0a/tests/test_wsimeta.py#L84

File: tests/test_wsi_registration.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    fixed_img = np.asarray(Image.open("sample.png"))
+    urllib.request.urlretrieve(main_url + "HE_2_level8_gray.png", "sample.png")
```
Comment (shaneahmed) [inline tests/test_wsi_registration.py:15]: Please use remote_sample in `conftest.py`.

File: tests/test_wsi_registration.py (hunk 3 above)
Comment (shaneahmed) [reply tests/test_wsi_registration.py:34]: This looks fine but it means the test on line 34 is redundant as it is covered in the code line 46 to 55.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 2 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:13]: Let's add these metrics once you have the final results.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    def __init__(self):
+        self.patch_size = (224, 224)
+        self.Xscale, self.Yscale = [], []
+        model = torchvision.models.vgg16(True)
+        return_layers = {"16": "block3_pool", "23": "block4_pool", "30": "block5_pool"}
+        self.FeatureExtractor = IntermediateLayerGetter(
+            model.features, return_layers=return_layers
+        )
+
+    def extract_features(self, fixed_img, moving_img):
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:26]: Does this function needs to be exposed? It can be made private when the full pipeline is implemented. At this stage making it private will make it difficult to run tests.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    def __init__(self):
+        self.patch_size = (224, 224)
+        self.Xscale, self.Yscale = [], []
+        model = torchvision.models.vgg16(True)
+        return_layers = {"16": "block3_pool", "23": "block4_pool", "30": "block5_pool"}
+        self.FeatureExtractor = IntermediateLayerGetter(
+            model.features, return_layers=return_layers
+        )
+
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:25]:
```suggestion
    # Make this function private when full pipeline is implemented.
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            raise ValueError(
+                "The required shape for fixed and moving images is n x m x 3."
+            )
+
+        if fixed_img.shape[2] != 3 or moving_img.shape[2] != 3:
+            raise ValueError("The input images are expected to have 3 channels.")
+
+        self.Xscale = 1.0 * np.array(fixed_img.shape[:2]) / self.patch_size
+        self.Yscale = 1.0 * np.array(moving_img.shape[:2]) / self.patch_size
+        fixed_cnn = cv2.resize(fixed_img, self.patch_size)
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:54]: use imresize in `tiatoolbox.utils.transforms`

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        moving_cnn = cv2.resize(moving_img, self.patch_size)
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:55]: Use https://github.com/TissueImageAnalytics/tiatoolbox/blob/af7319ef633e2f8b315b0949ed343f67d38bab0a/tiatoolbox/utils/transforms.py#L55

File: tests/test_wsi_registration.py (hunk 3 above)
Comment (ruqayya) [reply tests/test_wsi_registration.py]: The redundant code is removed now.

File: tests/test_wsi_registration.py (hunk 6 above)
Comment (ruqayya) [reply tests/test_wsi_registration.py]: Done.

File: tests/test_wsi_registration.py (hunk 7 above)
Comment (ruqayya) [reply tests/test_wsi_registration.py]: Added to remote_samples files:
 https://github.com/TissueImageAnalytics/tiatoolbox/blob/e89c666bf952ab1f0854cf1c5e7679e48b614e2d/tiatoolbox/data/remote_samples.yaml#L112-L115

File: tests/test_wsi_registration.py (hunk 8 above)
Comment (ruqayya) [reply tests/test_wsi_registration.py]: Added to remote_samples files:
 https://github.com/TissueImageAnalytics/tiatoolbox/blob/e89c666bf952ab1f0854cf1c5e7679e48b614e2d/tiatoolbox/data/remote_samples.yaml#L112-L115

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py:27]: Yes, we can make it private later.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 10 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py:26]: OK

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 11 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/e89c666bf952ab1f0854cf1c5e7679e48b614e2d/tiatoolbox/tools/registration/wsi_registration.py#L55-L56

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 12 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/e89c666bf952ab1f0854cf1c5e7679e48b614e2d/tiatoolbox/tools/registration/wsi_registration.py#L55-L56

File: tiatoolbox/data/remote_samples.yaml [hunk 13]
```diff
@@ -109,3 +109,9 @@ files:
     url: [*modelroot, "predictions/tissue_mask/small_svs_tissue_mask.npy"]
   micronet-output:
     url: [*modelroot, "predictions/nuclei_mask/micronet-output.npy"]
+  fixed_image:
+    url: [*testdata, "registration/HE_1_level8_gray.png"]
+  moving_image:
+    url: [ *testdata, "registration/HE_2_level8_gray.png" ]
```
Comment (John-P) [inline tiatoolbox/data/remote_samples.yaml:115]: These could maybe be JPEG files to make download for testing faster.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 10 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:27]: Please put this as a comment. It can be removed once this has been done.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:28]: Thanks. Please add the suggestion above and put the comment so this can be followed up later.

Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py:29]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/b718c77ef780c6930b720f3b8a37e130aea1f6b4/tiatoolbox/tools/registration/wsi_registration.py#L28-L29

File: tiatoolbox/data/remote_samples.yaml (hunk 13 above)
Comment (ruqayya) [reply tiatoolbox/data/remote_samples.yaml]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/b718c77ef780c6930b720f3b8a37e130aea1f6b4/tiatoolbox/data/remote_samples.yaml#L112-L115

File: tiatoolbox/data/remote_samples.yaml [hunk 14]
```diff
@@ -109,3 +109,9 @@ files:
     url: [*modelroot, "predictions/tissue_mask/small_svs_tissue_mask.npy"]
   micronet-output:
     url: [*modelroot, "predictions/nuclei_mask/micronet-output.npy"]
+  fixed_image:
+    url: [*testdata, "registration/HE_1_level8_gray.jpeg"]
+  moving_image:
+    url: [ *testdata, "registration/HE_2_level8_gray.jpeg" ]
+  dfbr_features:
+    url: [ *testdata, "registration/features.npy" ]
```
Comment (shaneahmed) [inline tiatoolbox/data/remote_samples.yaml:117]:
```suggestion
    url: [ *testdata, "registration/dfbr_features.npy" ]
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 15]
```diff
@@ -1,5 +1,78 @@
...
+import torch
+import torchvision
 from skimage import exposure, filters, morphology
+from torchvision.models._utils import IntermediateLayerGetter
+
+from tiatoolbox.utils.transforms import imresize
+
+
+class DFBRegistrtation:
+    r"""Deep Feature based Registration
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:11]:
```suggestion
    r"""Defines Deep Feature based Registration.
```

**Review by shaneahmed — APPROVED** [review]
Thanks @ruqayya
Looks good to me.

File: tests/conftest.py [hunk 16]
```diff
@@ -393,6 +393,14 @@ def moving_image(remote_sample) -> pathlib.Path:
     return remote_sample("moving_image")
 
 
+@pytest.fixture(scope="session")
+def dfbr_features(remote_sample) -> pathlib.Path:
+    """Sample pytest fixture for DFBR features.
+    Download features for pytest.
+    """
+    return remote_sample("dfbr_features")
```
Comment (John-P) [inline tests/conftest.py:401]: Prefer a descriptive name to acronyms, i.e. what is 'dfbr'. At least explain in the docstring.

File: tests/conftest.py [hunk 17]
```diff
@@ -393,6 +393,14 @@ def moving_image(remote_sample) -> pathlib.Path:
     return remote_sample("moving_image")
 
 
+@pytest.fixture(scope="session")
+def dfbr_features(remote_sample) -> pathlib.Path:
+    """Sample pytest fixture for DFBR features.
+    Download features for pytest.
```
Comment (John-P) [inline tests/conftest.py:399]:
```suggestion
    """Sample pytest fixture for DFBR features.
    
    Download features for pytest.
    
```

File: tests/conftest.py [hunk 18]
```diff
@@ -393,6 +393,14 @@ def moving_image(remote_sample) -> pathlib.Path:
...
+    """
+    return remote_sample("dfbr_features")
+
+
 @pytest.fixture(scope="session")
 def fixed_mask(remote_sample) -> pathlib.Path:
     """Sample pytest fixture for fixed mask.
```
Comment (John-P) [inline tests/conftest.py:406]: Docstring new lines missing here.

**Review by John-P — COMMENTED** [review]
Looks good, just a couple of minor comments. It would be nice to add type hints also. I can help with this is you like

**Comment by John-P** [comment]
One general comment is the name of the class. It is named functionally e.g. "Registration" rather than as an object e.g. "Register" i.e. 'the thing that does the registering'. I would suggest changing the naming to make it more clear what is an object / function.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 19]
```diff
@@ -172,3 +176,71 @@ def match_histograms(image_a, image_b, kernel_size=7):
         image_a = exposure.match_histograms(image_a, image_b).astype(np.uint8)
 
     return image_a, image_b
+
+
+class DFBRegistrtation:
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:181]:
```suggestion
class DFBRegister:
```

File: tests/conftest.py (hunk 16 above)
Comment (ruqayya) [reply tests/conftest.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/4a4290bd4a06835ac1f8e333f4396e764e0555a8/tests/conftest.py#L399-L404

File: tests/conftest.py (hunk 18 above)
Comment (ruqayya) [reply tests/conftest.py:410]: Added in other places as well. 
https://github.com/TissueImageAnalytics/tiatoolbox/blob/4a4290bd4a06835ac1f8e333f4396e764e0555a8/tests/conftest.py#L381-L386
